ssl/s3_lib.c

```diff
@@ -4374,11 +4406,17 @@ int ssl3_get_req_cert_type(SSL *s, WPACKET *pkt)
 
 #ifndef OPENSSL_NO_GOST
     if (s->version >= TLS1_VERSION && (alg_k & SSL_kGOST))
-            return WPACKET_put_bytes_u8(pkt, TLS_CT_GOST01_SIGN)
-                    && WPACKET_put_bytes_u8(pkt, TLS_CT_GOST12_IANA_SIGN)
-                    && WPACKET_put_bytes_u8(pkt, TLS_CT_GOST12_IANA_512_SIGN)
-                    && WPACKET_put_bytes_u8(pkt, TLS_CT_GOST12_LEGACY_SIGN)
-                    && WPACKET_put_bytes_u8(pkt, TLS_CT_GOST12_LEGACY_512_SIGN);
+        if (WPACKET_put_bytes_u8(pkt, TLS_CT_GOST01_SIGN) == 0
+            || WPACKET_put_bytes_u8(pkt, TLS_CT_GOST12_IANA_SIGN) == 0
+            || WPACKET_put_bytes_u8(pkt, TLS_CT_GOST12_IANA_512_SIGN) == 0
+            || WPACKET_put_bytes_u8(pkt, TLS_CT_GOST12_LEGACY_SIGN) == 0
+            || WPACKET_put_bytes_u8(pkt, TLS_CT_GOST12_LEGACY_512_SIGN == 0))
+            return 0;
+
+    if (s->version >= TLS1_2_VERSION && (alg_k & SSL_kGOST18))
+        if (WPACKET_put_bytes_u8(pkt, TLS_CT_GOST12_IANA_SIGN) == 0
+            || WPACKET_put_bytes_u8(pkt, TLS_CT_GOST12_IANA_512_SIGN) == 0)
```

Comment:
Style nit: everywhere else we do `!WPACKET...` instead of `WPACKET... == 0`

Reply:
Fixed. Thanks!

Reply:
I've tested it against an alternative implementation, and it works fine. I'll try to add at least self-compatibility tests to the gost-engine test suite.